server/engine/src/main/java/org/eclipse/lsp/cobol/core/visitor/CobolVisitor.java

```diff
@@ -90,27 +88,33 @@ public class CobolVisitor extends CobolParserBaseVisitor<List<Node>> {
   private final Map<String, SubroutineDefinition> subroutineDefinitionMap = new HashMap<>();
   protected final CachingConfigurationService cachingConfigurationService;
 
+  private final TextExtractionState text;
+
   public CobolVisitor(
           @NonNull CopybooksRepository copybooks,
           @NonNull CommonTokenStream tokenStream,
           @NonNull ExtendedDocument extendedDocument,
           MessageService messageService,
           SubroutineService subroutineService,
-          CachingConfigurationService cachingConfigurationService, CobolProgramLayout programLayout) {
+          CachingConfigurationService cachingConfigurationService,
+          CobolProgramLayout programLayout) {
     this.copybooks = copybooks;
     this.tokenStream = tokenStream;
     this.extendedDocument = extendedDocument;
     this.messageService = messageService;
     this.subroutineService = subroutineService;
     this.cachingConfigurationService = cachingConfigurationService;
     this.programLayout = programLayout;
+    this.text = new TextExtractionState(extendedDocument);
   }
 
   @Override
   public List<Node> visitStartRule(StartRuleContext ctx) {
     // we can skip the other nodes, but not the root
-    return ImmutableList.of(
-            retrieveLocality(ctx)
+    text.setInputStream(ctx.getStart().getInputStream());
+    try {
+      return ImmutableList.of(
+            retrieveLocality(ctx, extendedDocument, copybooks)
                     .map(RootNode::new)
                     .map(
                             rootNode -> {
```

Comment:
should a flush/reset be here as well?

Reply:
May be.. Do you have a case in mind?

Reply:
nested program units.

Reply:
It's the start rule, it should be the only one for the CU. And as nested programs must be at the end of procedure division body, we should be ok in the good cases.